src/ec/system76/ec/system76_ec.c

```diff
@@ -118,33 +123,529 @@ uint8_t system76_ec_read_version(uint8_t *data)
+		if ((rv = ec_spi_erase_sector(addr)))
+			return -rv;
+	}
+
+	return addr;
+}
+
+/*
+ * Program an entire chip with a given image.
+ * Returns written byte count on success and <0 on error.
+ */
+static long ec_spi_image_write(uint8_t *image, ssize_t size)
+{
+	int status;
+	uint8_t rv;
+	uint32_t addr;
+	uint8_t buf[6] = {0};
+
+	if ((rv = ec_spi_cmd_write_enable()))
+		return -rv;
+
+	/* SPI AAI Word Program */
+	buf[0] = 0xAD;
+
+	for (addr = 0; addr < size; addr += 2) {
+		if ((rv = ec_spi_reset()))
+			return -rv;
+
+		if (addr == 0) {
+			/* 1st cmd bytes 1,2,3 are the address */
+			buf[4] = image[0];
+			buf[5] = image[1];
+
+			rv = ec_spi_bus_write(buf, 6);
+		} else {
+			buf[1] = image[addr];
+			buf[2] = image[addr + 1];
+
+			rv = ec_spi_bus_write(buf, 3);
+		}
+
+		if (rv)
+			return -rv;
+
+		do {
+			status = ec_spi_cmd_status();
+		} while (status > 0 && (status & 1) != 0);
+	}
+
+	if ((rv = ec_spi_cmd_write_disable()))
+		return -rv;
+
+	return addr;
+}
+
+/* Read a sector into dest. Returns 0 on success and <0 on error. */
+static long ec_spi_read_sector(uint8_t *dest, uint32_t addr)
+{
+	uint8_t rv;
+	uint8_t buf[5] = {0};
+
+	/* SPI Read */
+	buf[0] = 0x0B;
+	buf[1] = (addr >> 16) & 0xFF;
+	buf[2] = (addr >> 8) & 0xFF;
+	buf[3] = addr & 0xFF;
+	buf[4] = 0;
+
+	if ((rv = ec_spi_reset()))
+		return -rv;
+
+	if ((rv = ec_spi_bus_write(buf, 5)))
+		return -rv;
+
+	if ((rv = ec_spi_bus_read(dest, SPI_SECTOR_SIZE)))
+		return -rv;
+
+	return 0;
+}
+
+/* Verify an image sector by sector. Returns 0 on success and <0 on error. */
+static long ec_spi_image_verify(uint8_t *image, ssize_t image_sz)
+{
+	uint8_t *sector;
+	uint32_t addr;
+	long rv;
+
+	sector = malloc(SPI_SECTOR_SIZE);
+
+	for (addr = 0; addr < CONFIG_EC_SYSTEM76_EC_FLASH_SIZE && addr + SPI_SECTOR_SIZE < image_sz;
+	     addr += SPI_SECTOR_SIZE) {
+		if ((rv = ec_spi_read_sector(sector, addr)))
+			return -rv;
+
+		if (memcmp(sector, image + addr, SPI_SECTOR_SIZE))
+			return -1;
+	}
+
+	if (addr == image_sz)
+		return 0;
+
+	if ((rv = ec_spi_read_sector(sector, addr)))
+		return -rv;
+
+	if (memcmp(sector, image + addr, image_sz % SPI_SECTOR_SIZE))
+		return -1;
+
+	return 0;
 }
+
+/* Sync the EC firmware with an image contained in CBFS. */
+static void system76_ec_fw_sync(void *unused)
+{
+	size_t image_sz;
+	char *image;
+	char img_board_str[64];
+	char img_version_str[64];
+	char cur_board_str[64];
+	char cur_version_str[64];
+	uint8_t smfi_cmd;
+	long rv;
+
+	if (vboot_recovery_mode_enabled()) {
+		printk(BIOS_DEBUG, "Skipping EC update in recovery mode.\n");
+		return;
+	}
+
+	image = cbfs_map("ec.rom", &image_sz);
+	if (image == NULL) {
+		printk(BIOS_DEBUG, "No EC firmware update found.\n");
+		return;
+	}
+
+	printk(BIOS_DEBUG, "EC update found (%ld bytes)\n", image_sz);
+
+	/* Sanity checks */
+	if ((ec_read(0x10) & 0x01) != 0x01) {
+		printk(BIOS_WARNING, "AC adapter not connected. Skipping update.\n");
+		return;
+	}
```

Comment:
Ideally if we could display something on the screen... Booting with unmatched EC version may be risky anyways.

I wonder if it would make more sense to do it in UEFI payload with more rich graphical interface.

Reply:
Or use the new SoL feature for Alderlake to display text messages

Reply:
Works as expected, goes away after a reboot with an AC adapter connected. https://github.com/Dasharo/edk2/pull/127/files

Reply:
I can already hear compliants about yet another scary pop-up...

To me it looks good and I like it of course. I prefer being informed about problems.